**Importación librerias necesarías para realizar el EDA.**

In [2]:
# Tratamiento de datos
import pandas as pd
import numpy as np

**1. EDA del primer dataset ('bank-additional.csv')**:

In [17]:
# Cargar el archivo csv 'bank-additional.csv':
bank = pd.read_csv('Data/bank-additional.csv',index_col = 0) # index_col = 0 para omitir el indice que pone pandas al cargar el datraframe

# Visualizar nombre de las columnas y las primeras 5 filas para comprobar que se ha cargado correctamente:
bank.head() # 5 primeras filas

age        job  marital    education  default  housing  loan    contact  \
0   NaN  housemaid  MARRIED     basic.4y      0.0      0.0   0.0  telephone   
1  57.0   services  MARRIED  high.school      NaN      0.0   0.0  telephone   
2  37.0   services  MARRIED  high.school      0.0      1.0   0.0  telephone   
3  40.0     admin.  MARRIED     basic.6y      0.0      0.0   0.0  telephone   
4  56.0   services  MARRIED  high.school      0.0      0.0   1.0  telephone   

   duration  campaign  ...  emp.var.rate  cons.price.idx cons.conf.idx  \
0       261         1  ...           1.1          93,994         -36,4   
1       149         1  ...           1.1          93,994         -36,4   
2       226         1  ...           1.1          93,994         -36,4   
3       151         1  ...           1.1          93,994         -36,4   
4       307         1  ...           1.1          93,994         -36,4   

   euribor3m nr.employed   y                date latitude longitude  \
0      4,857        5191  no       2-agosto-2019   41.495   -71.233   
1        NaN        5191  no  14-septiembre-2016   34.601   -83.923   
2      4,857        5191  no     15-febrero-2019   34.939   -94.847   
3        NaN        5191  no   29-noviembre-2015   49.041   -70.308   
4        NaN        5191  no       29-enero-2017   38.033  -104.463   

                                    id_  
0  089b39d8-e4d0-461b-87d4-814d71e0e079  
1  e9d37224-cb6f-4942-98d7-46672963d097  
2  3f9f49b5-e410-4948-bf6e-f9244f04918b  
3  9991fafb-4447-451a-8be2-b0df6098d13e  
4  eca60b76-70b6-4077-80ba-bc52e8ebb0eb  

[5 rows x 23 columns]

In [5]:
# Dimensiones del set de datos:
print(f'Número de filas: {bank.shape[0]}\nNúmero de columnas: {bank.shape[1]}') 

Número de filas: 43000
Número de columnas: 23


In [7]:
# Visualizar el nombre de las columnas con el tipo de datos y si tienen nulos.
bank.info()

<class 'pandas.core.frame.DataFrame'>
Index: 43000 entries, 0 to 27570
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             37880 non-null  float64
 1   job             42655 non-null  object 
 2   marital         42915 non-null  object 
 3   education       41193 non-null  object 
 4   default         34019 non-null  float64
 5   housing         41974 non-null  float64
 6   loan            41974 non-null  float64
 7   contact         43000 non-null  object 
 8   duration        43000 non-null  int64  
 9   campaign        43000 non-null  int64  
 10  pdays           43000 non-null  int64  
 11  previous        43000 non-null  int64  
 12  poutcome        43000 non-null  object 
 13  emp.var.rate    43000 non-null  float64
 14  cons.price.idx  42529 non-null  object 
 15  cons.conf.idx   43000 non-null  object 
 16  euribor3m       33744 non-null  object 
 17  nr.employed     43000 non-null  obje

**Columnas con nulos**: age, job, marital, education, default, housing, loan, cons.price.idx, euribor3m, date


**Columnas con tipo de datos erroneo**:

cons.price.idx (texto) convertir a decimal
cons.conf.idx (texto) conveertir a decimal
euribor3m (texto) convertir a decimal
nr.employed (texto) convertira entero
date (texto) convertir a fecha
default (decimal) convertir a entero
housing (decimal) convertir a entero
loan (decimal) convertir a entero

**Descripción estadistica de las columnas numéricas**:

In [8]:
bank.describe().T # descripción estadistica de columnas numéricas por defecto, con .T muestra las transposición de las mismas

count        mean         std      min        25%       50%  \
age           37880.0   39.977112   10.437957   17.000   32.00000   38.0000   
default       34019.0    0.000088    0.009390    0.000    0.00000    0.0000   
housing       41974.0    0.535998    0.498708    0.000    0.00000    1.0000   
loan          41974.0    0.155620    0.362499    0.000    0.00000    0.0000   
duration      43000.0  257.739279  258.666033    0.000  102.00000  179.0000   
campaign      43000.0    2.567233    2.772294    1.000    1.00000    2.0000   
pdays         43000.0  962.330953  187.260394    0.000  999.00000  999.0000   
previous      43000.0    0.174023    0.497366    0.000    0.00000    0.0000   
emp.var.rate  43000.0    0.077128    1.573898   -3.400   -1.80000    1.1000   
latitude      43000.0   36.856697    7.225948   24.396   30.61475   36.7610   
longitude     43000.0  -95.939067   16.752282 -124.997 -110.49425  -95.8995   

                    75%       max  
age            47.00000    98.000  
default         0.00000     1.000  
housing         1.00000     1.000  
loan            0.00000     1.000  
duration      319.00000  4918.000  
campaign        3.00000    56.000  
pdays         999.00000   999.000  
previous        0.00000     7.000  
emp.var.rate    1.40000     1.400  
latitude       43.11325    49.384  
longitude     -81.42775   -66.937

La información de este dataframe muestra:

- default, housing y loan tienen unos y ceros
- pdays tiene una anomalia en los valores teniendo 999 en los días que han pasado desde el último contacto, posiblemente sea el valor arbitrario que le daban en el banco a los clientes con los que no se había contactado
- duration puede que tenga outliers
- campaign puede que tenga outliers
- latitude y longitude son coordenadas y no es necesario sacar información estadistica de ellas

**Descripción estadistica de las columnas categóricas**:

In [9]:
bank.describe(include ='O').T # para las columnas categóricas, la "O" equivale a object

count unique                                   top   freq
job             42655     11                                admin.  10873
marital         42915      3                               MARRIED  25999
education       41193      7                     university.degree  12722
contact         43000      2                              cellular  27396
poutcome        43000      3                           NONEXISTENT  37103
cons.price.idx  42529     26                                93,994   7938
cons.conf.idx   43000     26                                 -36,4   8020
euribor3m       33744    309                                 4,857   2287
nr.employed     43000     11                                5228,1  16980
y               43000      2                                    no  38156
date            42752   1825                       28-febrero-2018     95
id_             43000  43000  993bbbd6-4dbc-4a40-a408-f91f8462bee6      1

id_ no se repite, por lo que es el identificador único de cada cliente.

**Elementos únicos de las columnas categóricas y la cantidad de veces que se repiten**:

In [10]:
# Primero un Data Frame sólo con las columnas categóricas
bank_categoricas = bank.select_dtypes(include = 'O')

# Lista con los nombres de todas las columnas
lista_columnas = bank_categoricas.columns.tolist()# Al añadir .tolist() convierte el tipo index a lista para poder iterar por él.

# Para ver los valores únicos y el conteo de cada uno
for columna in lista_columnas:
    print(f'{columna.upper()}\n')
    print(f'Valóres únicos: {bank[columna].unique()}\n')# Imprime los valores únicos de la columna
    print(f'Columna {bank[columna].value_counts()}') # Contador de cada valor único
    print('----------------------------------------------------\n')

JOB

Valóres únicos: ['housemaid' 'services' 'admin.' 'blue-collar' 'technician' 'retired'
 'management' 'unemployed' 'self-employed' nan 'entrepreneur' 'student']

Columna job
admin.           10873
blue-collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self-employed     1489
housemaid         1123
unemployed        1063
student            903
Name: count, dtype: int64
----------------------------------------------------

MARITAL

Valóres únicos: ['MARRIED' 'SINGLE' 'DIVORCED' nan]

Columna marital
MARRIED     25999
SINGLE      12105
DIVORCED     4811
Name: count, dtype: int64
----------------------------------------------------

EDUCATION

Valóres únicos: ['basic.4y' 'high.school' 'basic.6y' 'basic.9y' 'professional.course' nan
 'university.degree' 'illiterate']

Columna education
university.degree      12722
high.school             9925
basic.9y                6309
professional.course     5477
basic.4

**Análisis de los valores de las columnas categóricas**:

job: Valor más repetido 'admin', 11 únicos y tiene nulos.

marital: Valores únicos 3 (MARRIED, SINGLE, DIVORCED), más repetido 'MARRIED' y tiene nuloseducation:7 únicos. Valor más repetido 'university degree', menos repetido 'illiterate' con 18 registros. Tiene nulos

contact: 2 valores (telephone y cellular). El segundo tiene casi el doble de registros que el primero. No tiene nulos

poutcome: 3 valores (NONEXISTENT, FAILURE, SUCCESS). Tenemos notablemente más datos del primero que del resto. No tiene nulos

cons.price.idx: Es una falsa columna categórica. Tiene nulos

cons.conf.idx: Es una falsa columna categórica. No tiene nulos

euribor3m: Es una falsa columna categórica.Tiene nulos

nr.employed: Es una falsa columna categórica. No tiene nulos

y: 2 valores (no, yes). No: 38156, Yes: 4844. No tiene nulos.

date: Falsa columna categórica.es una fecha

id_: Identificador único. Sin nulos.

**Cantidad de nulos por columna**:

In [11]:
print('Nulos por columna:')
bank.isnull().sum()

Nulos por columna:


age               5120
job                345
marital             85
education         1807
default           8981
housing           1026
loan              1026
contact              0
duration             0
campaign             0
pdays                0
previous             0
poutcome             0
emp.var.rate         0
cons.price.idx     471
cons.conf.idx        0
euribor3m         9256
nr.employed          0
y                    0
date               248
latitude             0
longitude            0
id_                  0
dtype: int64

**Cantidad de valores duplicados:**

In [13]:
print('Cantidad de duplicados por columna:')
bank.duplicated().sum() 

Cantidad de duplicados por columna:


np.int64(0)

**2. EDA del segundo Dataset 'customer-details.xlsx'**:

Data Frame para cada hoja del archivo de excel:

In [20]:
#DF distintos, uno para cada hoja.
customer_2012 = pd.read_excel('Data/customer-details.xlsx', sheet_name='2012', index_col = 0)
customer_2013 = pd.read_excel('Data/customer-details.xlsx', sheet_name='2013', index_col = 0)
customer_2014 = pd.read_excel('Data/customer-details.xlsx', sheet_name='2014', index_col = 0)

# Lista con los 3 DataFrames para poder hacer el EDA de los 3 al mismo tiempo
lista_customer = [customer_2012,customer_2013,customer_2014]

# Ver primeras filas de cada DF
for df in lista_customer:
    display(df.head())

Income  Kidhome  Teenhome Dt_Customer  NumWebVisitsMonth  \
0  161770        1         0  2012-04-04                 29   
1   85477        1         1  2012-12-30                  7   
2  147233        1         1  2012-02-02                  5   
3  121393        1         2  2012-12-21                 29   
4   63164        1         2  2012-06-20                 20   

                                     ID  
0  089b39d8-e4d0-461b-87d4-814d71e0e079  
1  e9d37224-cb6f-4942-98d7-46672963d097  
2  3f9f49b5-e410-4948-bf6e-f9244f04918b  
3  9991fafb-4447-451a-8be2-b0df6098d13e  
4  eca60b76-70b6-4077-80ba-bc52e8ebb0eb

Income  Kidhome  Teenhome Dt_Customer  NumWebVisitsMonth  \
0   82407        0         1  2013-10-07                 29   
1   24877        0         0  2013-04-21                 32   
2   78839        1         0  2013-11-25                 21   
3  113859        1         1  2013-01-06                 12   
4  147317        0         0  2013-07-10                 14   

                                     ID  
0  ef81336c-e41a-46d0-8a30-5d4ac3b836be  
1  50b224ac-84b4-42a7-aba9-3e1d1d620479  
2  9cb15c53-b647-464a-add9-09cdae04897c  
3  f8aaf334-949d-4943-a214-b64fe9c088b4  
4  885fe85a-3718-4760-a32a-a57628cf7528

Income  Kidhome  Teenhome Dt_Customer  NumWebVisitsMonth  \
0  168812        1         1  2014-06-10                 27   
1  108615        2         1  2014-04-05                 24   
2  179041        0         2  2014-09-22                  4   
3  116639        2         2  2014-08-16                  3   
4   49537        0         1  2014-09-05                 28   

                                     ID  
0  4d6c811c-0f24-4a73-b374-2ebd1f80d46e  
1  eef0f24f-cbdb-4c67-b371-09d7bc59695e  
2  bbfdcedd-199d-4927-94bb-5dad051ff471  
3  e8ebb67d-4ee1-4615-b701-eefe3af607eb  
4  f7f79ace-0024-4505-a401-f340946f0942

Tienen las mismas columnas con los mismos nombres.

**Dimensiones de cada uno de los data frames.**

In [23]:
# Iterar en un enumerate con la lista_customer y empieza en 2012
for year, df in enumerate(lista_customer, start = 2012):
    print(f'Número de filas  customer_{year}: {df.shape[0]} - Número de columnas: {df.shape[1]}') #tupla con filas y columnas.

print('----------------------\n')
print('Información de los 3 DataFrames:\n')
for df in lista_customer:
    print(df.info())

Número de filas  customer_2012: 20115 - Número de columnas: 6
Número de filas  customer_2013: 8965 - Número de columnas: 6
Número de filas  customer_2014: 14090 - Número de columnas: 6
----------------------

Información de los 3 DataFrames:

<class 'pandas.core.frame.DataFrame'>
Index: 20115 entries, 0 to 20114
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Income             20115 non-null  int64         
 1   Kidhome            20115 non-null  int64         
 2   Teenhome           20115 non-null  int64         
 3   Dt_Customer        20115 non-null  datetime64[ns]
 4   NumWebVisitsMonth  20115 non-null  int64         
 5   ID                 20115 non-null  object        
dtypes: datetime64[ns](1), int64(4), object(1)
memory usage: 1.1+ MB
None
<class 'pandas.core.frame.DataFrame'>
Index: 8965 entries, 0 to 8964
Data columns (total 6 columns):
 #   Column             Non-Null Count 

Una vez comprobado  la cantidad de columnas y que son del mismo tipo, se procede a  hacer la unión de los data frames en uno solo.

**Unir verticalmente los data frames con las hojas de Excel:**

In [28]:
# Concatenar los data frames indexado por el indice único
customer_concat = pd.concat([customer_2012, customer_2013, customer_2014], axis = 0, ignore_index = True)
customer_concat.head() # muestra de las primeras filas



Income  Kidhome  Teenhome Dt_Customer  NumWebVisitsMonth  \
0  161770        1         0  2012-04-04                 29   
1   85477        1         1  2012-12-30                  7   
2  147233        1         1  2012-02-02                  5   
3  121393        1         2  2012-12-21                 29   
4   63164        1         2  2012-06-20                 20   

                                     ID  
0  089b39d8-e4d0-461b-87d4-814d71e0e079  
1  e9d37224-cb6f-4942-98d7-46672963d097  
2  3f9f49b5-e410-4948-bf6e-f9244f04918b  
3  9991fafb-4447-451a-8be2-b0df6098d13e  
4  eca60b76-70b6-4077-80ba-bc52e8ebb0eb

In [27]:
print(f'Número de filas en customer_concat: {customer_concat.shape[0]} - Número de columnas: {customer_concat.shape[1]}')   #Comprobar las columnas y si el número de filas coincide con las suma de las filas de los data frames concatenados.

Número de filas en customer_concat: 43170 - Número de columnas: 6


In [29]:
customer_concat.info() # información de las columnas del nuevo dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43170 entries, 0 to 43169
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Income             43170 non-null  int64         
 1   Kidhome            43170 non-null  int64         
 2   Teenhome           43170 non-null  int64         
 3   Dt_Customer        43170 non-null  datetime64[ns]
 4   NumWebVisitsMonth  43170 non-null  int64         
 5   ID                 43170 non-null  object        
dtypes: datetime64[ns](1), int64(4), object(1)
memory usage: 2.0+ MB


In [30]:
print('Cantidad de nulos:')
customer_concat.isnull().sum() # suma de los nulos

Cantidad de nulos:


Income               0
Kidhome              0
Teenhome             0
Dt_Customer          0
NumWebVisitsMonth    0
ID                   0
dtype: int64

In [31]:
print('Cantidad de duplicados:', customer_concat.duplicated().sum()) # suma de los duplicados

Cantidad de duplicados: 0


**Comparación de las dimensiones de los dos dataframes, customer_concat y bank.**

In [32]:
print(f'Número de filas en bank: {bank.shape[0]}, y el número de columnas: {bank.shape[1]}')
print(f'Número de filas en customer_concat: {customer_concat.shape[0]}, y el número de columnas: {customer_concat.shape[1]}')

Número de filas en bank: 43000, y el número de columnas: 23
Número de filas en customer_concat: 43170, y el número de columnas: 6


In [34]:
bank.head(3)

age        job  marital    education  default  housing  loan    contact  \
0   NaN  housemaid  MARRIED     basic.4y      0.0      0.0   0.0  telephone   
1  57.0   services  MARRIED  high.school      NaN      0.0   0.0  telephone   
2  37.0   services  MARRIED  high.school      0.0      1.0   0.0  telephone   

   duration  campaign  ...  emp.var.rate  cons.price.idx cons.conf.idx  \
0       261         1  ...           1.1          93,994         -36,4   
1       149         1  ...           1.1          93,994         -36,4   
2       226         1  ...           1.1          93,994         -36,4   

   euribor3m nr.employed   y                date latitude longitude  \
0      4,857        5191  no       2-agosto-2019   41.495   -71.233   
1        NaN        5191  no  14-septiembre-2016   34.601   -83.923   
2      4,857        5191  no     15-febrero-2019   34.939   -94.847   

                                    id_  
0  089b39d8-e4d0-461b-87d4-814d71e0e079  
1  e9d37224-cb6f-4942-98d7-46672963d097  
2  3f9f49b5-e410-4948-bf6e-f9244f04918b  

[3 rows x 23 columns]

In [35]:
customer_concat.head()

Income  Kidhome  Teenhome Dt_Customer  NumWebVisitsMonth  \
0  161770        1         0  2012-04-04                 29   
1   85477        1         1  2012-12-30                  7   
2  147233        1         1  2012-02-02                  5   
3  121393        1         2  2012-12-21                 29   
4   63164        1         2  2012-06-20                 20   

                                     ID  
0  089b39d8-e4d0-461b-87d4-814d71e0e079  
1  e9d37224-cb6f-4942-98d7-46672963d097  
2  3f9f49b5-e410-4948-bf6e-f9244f04918b  
3  9991fafb-4447-451a-8be2-b0df6098d13e  
4  eca60b76-70b6-4077-80ba-bc52e8ebb0eb

# El identificador id es el campo en común por el que se puede hacer la unión de los dos dataframes.
# Hay mayor cantidad de clientes de los que han participado en la campaña.
# Unir los Dataframes horizontalmente para solo obtener los valores coincidentes:

In [36]:
bank = bank.merge(customer_concat,how = 'inner', left_on = 'id_', right_on = 'ID') # une los dos data frames por los valores coincidentes
bank.head()

age        job  marital    education  default  housing  loan    contact  \
0   NaN  housemaid  MARRIED     basic.4y      0.0      0.0   0.0  telephone   
1  57.0   services  MARRIED  high.school      NaN      0.0   0.0  telephone   
2  37.0   services  MARRIED  high.school      0.0      1.0   0.0  telephone   
3  40.0     admin.  MARRIED     basic.6y      0.0      0.0   0.0  telephone   
4  56.0   services  MARRIED  high.school      0.0      0.0   1.0  telephone   

   duration  campaign  ...                date  latitude longitude  \
0       261         1  ...       2-agosto-2019    41.495   -71.233   
1       149         1  ...  14-septiembre-2016    34.601   -83.923   
2       226         1  ...     15-febrero-2019    34.939   -94.847   
3       151         1  ...   29-noviembre-2015    49.041   -70.308   
4       307         1  ...       29-enero-2017    38.033  -104.463   

                                    id_  Income Kidhome Teenhome Dt_Customer  \
0  089b39d8-e4d0-461b-87d4-814d71e0e079  161770       1        0  2012-04-04   
1  e9d37224-cb6f-4942-98d7-46672963d097   85477       1        1  2012-12-30   
2  3f9f49b5-e410-4948-bf6e-f9244f04918b  147233       1        1  2012-02-02   
3  9991fafb-4447-451a-8be2-b0df6098d13e  121393       1        2  2012-12-21   
4  eca60b76-70b6-4077-80ba-bc52e8ebb0eb   63164       1        2  2012-06-20   

  NumWebVisitsMonth                                    ID  
0                29  089b39d8-e4d0-461b-87d4-814d71e0e079  
1                 7  e9d37224-cb6f-4942-98d7-46672963d097  
2                 5  3f9f49b5-e410-4948-bf6e-f9244f04918b  
3                29  9991fafb-4447-451a-8be2-b0df6098d13e  
4                20  eca60b76-70b6-4077-80ba-bc52e8ebb0eb  

[5 rows x 29 columns]

In [41]:
bank.drop('id_', axis = 1, inplace = True) # eliminar la columna id_ axis = 1 eliminar los datos de una columna, axis = 0 de una fila. inplace=True sobreescribir el DF

KeyError: "['id_'] not found in axis"

In [ ]:
bank.columns #Comprobar que la coluna se ha borrado correctamente:

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'duration', 'campaign', 'pdays', 'previous', 'poutcome',
       'emp.var.rate', 'cons.price.idx', 'cons.conf.idx', 'euribor3m',
       'nr.employed', 'y', 'date', 'latitude', 'longitude', 'Income',
       'Kidhome', 'Teenhome', 'Dt_Customer', 'NumWebVisitsMonth', 'ID'],
      dtype='object')

In [43]:
nuevas_columnas = {columna:columna.lower().replace('.', '_') for columna in bank.columns} # itera por los nombres de las columnas pasandolos a minusculas y sustiyuendo los . por _. Crea un diccionario cuya clave es el nombre antiguo y el valor es el nombre cambiado

In [44]:
bank.rename(columns = nuevas_columnas, inplace = True) # método rename para renombrar con los nombres del diccionario
bank.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'duration', 'campaign', 'pdays', 'previous', 'poutcome',
       'emp_var_rate', 'cons_price_idx', 'cons_conf_idx', 'euribor3m',
       'nr_employed', 'y', 'date', 'latitude', 'longitude', 'income',
       'kidhome', 'teenhome', 'dt_customer', 'numwebvisitsmonth', 'id'],
      dtype='object')

In [45]:
for columna in bank.columns.to_list(): # recorrer los valores de las columnas
    try:
        bank[columna] = bank[columna].str.lower() # pasar a minusculas
    except:
        pass

In [46]:
bank.head(3)

age        job  marital    education  default  housing  loan    contact  \
0   NaN  housemaid  married     basic.4y      0.0      0.0   0.0  telephone   
1  57.0   services  married  high.school      NaN      0.0   0.0  telephone   
2  37.0   services  married  high.school      0.0      1.0   0.0  telephone   

   duration  campaign  ...   y                date latitude  longitude  \
0       261         1  ...  no       2-agosto-2019   41.495    -71.233   
1       149         1  ...  no  14-septiembre-2016   34.601    -83.923   
2       226         1  ...  no     15-febrero-2019   34.939    -94.847   

   income kidhome teenhome dt_customer numwebvisitsmonth  \
0  161770       1        0  2012-04-04                29   
1   85477       1        1  2012-12-30                 7   
2  147233       1        1  2012-02-02                 5   

                                     id  
0  089b39d8-e4d0-461b-87d4-814d71e0e079  
1  e9d37224-cb6f-4942-98d7-46672963d097  
2  3f9f49b5-e410-4948-bf6e-f9244f04918b  

[3 rows x 28 columns]

In [47]:
bank.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 28 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   age                37880 non-null  float64       
 1   job                42655 non-null  object        
 2   marital            42915 non-null  object        
 3   education          41193 non-null  object        
 4   default            34019 non-null  float64       
 5   housing            41974 non-null  float64       
 6   loan               41974 non-null  float64       
 7   contact            43000 non-null  object        
 8   duration           43000 non-null  int64         
 9   campaign           43000 non-null  int64         
 10  pdays              43000 non-null  int64         
 11  previous           43000 non-null  int64         
 12  poutcome           43000 non-null  object        
 13  emp_var_rate       43000 non-null  float64       
 14  cons_p

In [48]:
fecha = bank['date'].str.split('-', expand = True) # genera un data frame con las columnas día, mes y año
fecha

0           1     2
0       2      agosto  2019
1      14  septiembre  2016
2      15     febrero  2019
3      29   noviembre  2015
4      29       enero  2017
...    ..         ...   ...
42995  13     octubre  2015
42996  17       marzo  2018
42997  15  septiembre  2016
42998  23  septiembre  2019
42999   6   noviembre  2019

[43000 rows x 3 columns]

In [49]:
# nueva columna mes
bank['contact_month'] = fecha[1]
# nueva columna año
bank['contact_year'] = fecha[2]

In [50]:
bank.head(2)

age        job  marital    education  default  housing  loan    contact  \
0   NaN  housemaid  married     basic.4y      0.0      0.0   0.0  telephone   
1  57.0   services  married  high.school      NaN      0.0   0.0  telephone   

   duration  campaign  ...  latitude  longitude  income  kidhome teenhome  \
0       261         1  ...    41.495    -71.233  161770        1        0   
1       149         1  ...    34.601    -83.923   85477        1        1   

  dt_customer numwebvisitsmonth                                    id  \
0  2012-04-04                29  089b39d8-e4d0-461b-87d4-814d71e0e079   
1  2012-12-30                 7  e9d37224-cb6f-4942-98d7-46672963d097   

  contact_month contact_year  
0        agosto         2019  
1    septiembre         2016  

[2 rows x 30 columns]

In [51]:
    #Diccionario con la clave nombre del mes y el valor el número del mes
meses = {'enero':'01',
        'febrero':'02',
        'marzo':'03',
        'abril':'04',
        'mayo':'05',
        'junio':'06',
        'julio':'07',
        'agosto':'08',
        'septiembre':'09',
        'octubre':'10',
        'noviembre':'11',
        'diciembre':'12'
    }
bank.replace({'date':meses}, regex = True,inplace = True) # reemplaza los nombres de los meses por el número

bank['date'] = pd.to_datetime(bank['date'], format = "%d-%m-%Y") # datetime para cambiar a tipo fecha

In [52]:
bank['date'].head()

0   2019-08-02
1   2016-09-14
2   2019-02-15
3   2015-11-29
4   2017-01-29
Name: date, dtype: datetime64[ns]

In [53]:
bank['education'] = bank['education'].str.replace('.',' ') # reemplaza los puntos por espacios

In [54]:
bank['education'].head()

0       basic 4y
1    high school
2    high school
3       basic 6y
4    high school
Name: education, dtype: object

In [ ]:
# Convertir a decimal las columnas: 'cons_price_idx','cons_conf_idx','euribor3m','nr_employed':
# lista con los nombres de las columnas a convertir
str_float = ['cons_price_idx','cons_conf_idx','euribor3m','nr_employed']
# Itera por la lista
for col in str_float:
    #Sustituye las ',' por '.'
    bank[col] = bank[col].str.replace(',','.')
    #función lambda para cambiar el tipo de dato
    bank[col] = bank[col].apply(lambda x: float(x))

In [56]:
bank.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 30 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   age                37880 non-null  float64       
 1   job                42655 non-null  object        
 2   marital            42915 non-null  object        
 3   education          41193 non-null  object        
 4   default            34019 non-null  float64       
 5   housing            41974 non-null  float64       
 6   loan               41974 non-null  float64       
 7   contact            43000 non-null  object        
 8   duration           43000 non-null  int64         
 9   campaign           43000 non-null  int64         
 10  pdays              43000 non-null  int64         
 11  previous           43000 non-null  int64         
 12  poutcome           43000 non-null  object        
 13  emp_var_rate       43000 non-null  float64       
 14  cons_p

In [57]:
# Cambiar los valores 0 y 1 por No y Si en las columnas: 'default','housing','loan':
# crear diccionario
diccionario_mapeo = {0: "No", 1: "Si"}
# lista con las columnas
mapear = ['default','housing','loan']
# Itera por la lista y cambia los valores
for col in mapear:
    bank[col] = bank[col].map(diccionario_mapeo)

In [59]:
bank['pdays'].head()

0    999
1    999
2    999
3    999
4    999
Name: pdays, dtype: int64

In [60]:
# Cambiar 999 por nan en la columna 'pdays':
# En la columna 'pdays' hay valores 999, lo que indica que ese valor se añadía cuando no se había contactado con el cliente por lo que lo aconsejable es cambiar ese valor por 'null' para que no altere el resultado si hay que hacer algún calculo con esta columna.

# método '.replace()' sobre la columna 'pdays' para reemplazar los valores 999 por nulos de numpy (np.nan)
bank['pdays'] = bank['pdays'].replace(999, np.nan)

In [62]:
bank['pdays'].head(3)

0   NaN
1   NaN
2   NaN
Name: pdays, dtype: float64

In [64]:
bank.to_csv('Data/bank-limpio.csv') # guardar data frame limpio en archivo csv

In [66]:
bank = pd.read_csv('data/bank-limpio.csv',index_col = 0) # Cargar el archivo con los datos limpios para continuar con la transformación.

In [67]:
# Sustituir los valores nulos en las columnas que los tengan, según cada necesidad individual.
# porcentaje de nulos en las columnas.
nulos = (bank.isnull().sum() / bank.shape[0]) * 100

# convierte el cálculo anterior a DataFrame
df_nulos = pd.DataFrame(nulos, columns = ['%_nulos'])

# Filtra las columnas que tengan nulos
df_nulos[df_nulos['%_nulos'] > 0]

%_nulos
age             11.906977
job              0.802326
marital          0.197674
education        4.202326
default         20.886047
housing          2.386047
loan             2.386047
pdays           96.306977
cons_price_idx   1.095349
euribor3m       21.525581
date             0.576744
contact_month    0.576744
contact_year     0.576744

In [68]:
# Gestión de nulos en las columnas categóricas:
#lista de columnas categóricas que tienen nulos
col_nulos_categoricas = bank[bank.columns[bank.isnull().any()]].select_dtypes(include = 'O').columns
print('Columnas categóricas con nulos: \n ')
print(col_nulos_categoricas)

Columnas categóricas con nulos: 
 
Index(['job', 'marital', 'education', 'default', 'housing', 'loan', 'date',
       'contact_month'],
      dtype='object')


In [69]:
# distribución de los valores por categoría
for col in col_nulos_categoricas:
    print(f'La distribución de las categorías para la columna {col.upper()}')
    print(f'{bank[col].value_counts() / bank.shape[0]*100} %') # % de aparición
    print('-----------------------------------------------------')

La distribución de las categorías para la columna JOB
job
admin.           25.286047
blue-collar      22.451163
technician       16.339535
services          9.679070
management        7.093023
retired           4.162791
entrepreneur      3.539535
self-employed     3.462791
housemaid         2.611628
unemployed        2.472093
student           2.100000
Name: count, dtype: float64 %
-----------------------------------------------------
La distribución de las categorías para la columna MARITAL
marital
married     60.462791
single      28.151163
divorced    11.188372
Name: count, dtype: float64 %
-----------------------------------------------------
La distribución de las categorías para la columna EDUCATION
education
university degree      29.586047
high school            23.081395
basic 9y               14.672093
professional course    12.737209
basic 4y               10.130233
basic 6y                5.548837
illiterate              0.041860
Name: count, dtype: float64 %
--------------

**Conclusiones de los valores únicos**
- job: 11 valores. El resultado principal es 'admin' 25%, seguido de 'blue-collar' 22% , no hay una diferencia significativa.

- marital: 3 valores (married,single y divorced), married destaca muy por encima del resto con un 60%.

- education: 7 valores. No hay ningún valor que destaque frente al resto.

- default: Sólo tenemos dos valores (No y Si). El 'No' tiene un 79% de aparición.

- housing: Valores 'Si' y 'No'. No hay ningún valor que destaque frente al otro.

- loan: Sólo tenemos dos valores (No y Si). El 'No' tiene un 82% de aparición

- date: Son fechas, los valores no destacan entre si.

- contact_month: Son meses y los valores se reparten de manera equitativa entre todos los meses.

In [72]:
print(col_nulos_categoricas)


Index(['job', 'marital', 'education', 'default', 'housing', 'loan', 'date',
       'contact_month'],
      dtype='object')


In [76]:
bank.job.valuecounts()

AttributeError: 'Series' object has no attribute 'valuecounts'